// @cloudscape-design/components/internal/plugins/controllers/drawers.js
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0
import debounce from '../../debounce';
export class DrawersController {
    constructor() {
        this.drawers = [];
        this.drawersRegistrationListener = null;
        this.scheduleUpdate = debounce(() => {
            var _a;
            (_a = this.drawersRegistrationListener) === null || _a === void 0 ? void 0 : _a.call(this, this.drawers);
        }, 0);
        this.registerDrawer = (config) => {
            this.drawers = this.drawers.concat(config);
            this.scheduleUpdate();
        };
        this.updateDrawer = (config) => {
            var _a;
            const { id: drawerId, resizable, badge, defaultSize } = config;
            const drawerIndex = this.drawers.findIndex(({ id }) => id === drawerId);
            const oldDrawerConfig = (_a = this.drawers) === null || _a === void 0 ? void 0 : _a[drawerIndex];
            if (drawerIndex >= 0 && oldDrawerConfig) {
                const drawers = this.drawers.slice();
                const drawerConfig = Object.assign({}, oldDrawerConfig);
                if (typeof resizable === 'boolean') {
                    drawerConfig.resizable = resizable;
                }
                if (typeof badge === 'boolean') {
                    drawerConfig.badge = badge;
                }
                if (typeof defaultSize === 'number') {
                    drawerConfig.defaultSize = defaultSize;
                }
                drawers[drawerIndex] = drawerConfig;
                this.drawers = drawers;
                this.scheduleUpdate();
            }
            else {
                throw new Error(`[AwsUi] [runtime drawers] drawer with id ${drawerId} not found`);
            }
        };
        this.onDrawersRegistered = (listener) => {
            if (this.drawersRegistrationListener !== null) {
                console.warn('[AwsUi] [runtime drawers] multiple app layout instances detected');
            }
            this.drawersRegistrationListener = listener;
            this.scheduleUpdate();
            return () => {
                this.drawersRegistrationListener = null;
            };
        };
        this.clearRegisteredDrawers = () => {
            this.drawers = [];
        };
    }
    installPublic(api = {}) {
        var _a, _b;
        (_a = api.registerDrawer) !== null && _a !== void 0 ? _a : (api.registerDrawer = this.registerDrawer);
        (_b = api.updateDrawer) !== null && _b !== void 0 ? _b : (api.updateDrawer = this.updateDrawer);
        return api;
    }
    installInternal(internalApi = {}) {
        var _a, _b;
        (_a = internalApi.clearRegisteredDrawers) !== null && _a !== void 0 ? _a : (internalApi.clearRegisteredDrawers = this.clearRegisteredDrawers);
        (_b = internalApi.onDrawersRegistered) !== null && _b !== void 0 ? _b : (internalApi.onDrawersRegistered = this.onDrawersRegistered);
        return internalApi;
    }
}
